packages/ui/src/Slider.tsx

```diff
@@ -0,0 +1,340 @@
+  const wrapperRef = useRef<HTMLDivElement>(null)
+  const firstRangeInputRef = useRef<HTMLInputElement>(null)
+  const secondRangeInputRef = useRef<HTMLInputElement>(null)
+
+  /**
+   * We use this callback to update value either as number or [number, number] based on a rangeGuard
+   * We also send an event that triggered the change
+   */
+  const updateValues = useCallback(
+    ([changedFirstValue, changedSecondValue]: [number | null, number | null], e: ChangeEvent<HTMLInputElement> | MouseEvent) => {
+      if (isRangeGuard(value, onChange)) {
+        onChange([changedFirstValue ?? firstValue, changedSecondValue ?? secondValue], e)
+      }
+      if (changedFirstValue !== null && isSingleValueGuard(value, onChange)) {
+        onChange(changedFirstValue, e)
+      }
+    },
+    [firstValue, secondValue, onChange, value],
+  )
+
+  /**
+   * The following code makes sure that the relevant thumb is moved to the position once
+   * user clicks on a track.
+   *
+   * 1. We always move the first thumb in a non-range mode.
+   * 2. Otherwise, we move the thumb that is closed to a click event.
+   * 3. We focus the thumb after it is clicked.
+   */
+  const clickOnTrackHandler = useCallback(
+    (e: MouseEvent) => {
+      const { offsetX, target } = e
+      if (disabled) {
+        return false
+      }
+      if (!wrapperRef.current || wrapperRef.current !== target) {
+        return
+      }
+      const clickPercentage = offsetX / wrapperRef.current.offsetWidth
+      const resultStepSpot: number = Math.round((clickPercentage * max) / step) * step
+      if (!isRange || Math.abs(firstValue - resultStepSpot) < Math.abs(secondValue - resultStepSpot)) {
+        updateValues([resultStepSpot, null], e)
+        firstRangeInputRef.current?.focus()
+      } else {
+        // otherwise, let's move the right one
+        updateValues([null, resultStepSpot], e)
+        secondRangeInputRef.current?.focus()
+      }
+    },
+    [wrapperRef, firstValue, secondValue, disabled, isRange, max, step, updateValues],
+  )
+
+  useEffect(() => {
+    window.addEventListener('click', clickOnTrackHandler)
+    return () => {
+      window.removeEventListener('click', clickOnTrackHandler)
+    }
+  }, [clickOnTrackHandler])
+
+  /**
+   * The following callbacks reacts to native HTML events and update the values
+   */
+  const setFirstValueCallback = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      const {
+        target: { value },
+      } = e
+      const newValue = parseInt(value, 10)
+      if (newValue > secondValue) {
+        return false
+      }
+      updateValues([newValue, null], e)
+    },
+    [secondValue, updateValues],
+  )
+
+  const setSecondValueCallback = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      const {
+        target: { value },
+      } = e
+      const newValue = parseInt(value, 10)
+      if (newValue < firstValue) {
+        return false
+      }
+      updateValues([null, newValue], e)
```

Comment:
nit: a bit more readable imo.
```suggestion
      if (newValue >= firstValue) {
        updateValues([null, newValue], e)
      }
```